lib/srv/authhandlers_test.go

```diff
@@ -0,0 +1,235 @@
+limitations under the License.
+*/
+
+package srv
+
+import (
+	"context"
+	"net"
+	"testing"
+
+	"github.com/gravitational/trace"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/crypto/ssh"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/utils/sshutils"
+	"github.com/gravitational/teleport/lib/auth/native"
+	"github.com/gravitational/teleport/lib/auth/testauthority"
+	"github.com/gravitational/teleport/lib/events/eventstest"
+	"github.com/gravitational/teleport/lib/services"
+)
+
+type mockCAGetter struct {
+	AccessPoint
+
+	cas map[types.CertAuthType]types.CertAuthority
+}
+
+func (m mockCAGetter) GetCertAuthorities(ctx context.Context, caType types.CertAuthType, loadKeys bool, opts ...services.MarshalOption) ([]types.CertAuthority, error) {
+	ca, ok := m.cas[caType]
+	if !ok {
+		return nil, trace.NotFound("CA not found")
+	}
+
+	return []types.CertAuthority{ca}, nil
+}
+
+type mockLoginChecker struct {
+	authWithoutRBACChecked bool
+	authWithRBACChecked    bool
+}
+
+func (m *mockLoginChecker) canLoginWithoutRBAC(_ *ssh.Certificate, _ string, _, _ string) error {
+	m.authWithoutRBACChecked = true
+	return nil
+}
+
+func (m *mockLoginChecker) canLoginWithRBAC(_ *ssh.Certificate, _ string, _ types.Server, _, _ string) error {
+	m.authWithRBACChecked = true
+	return nil
+}
+
+type mockConnMetadata struct{}
+
+func (m mockConnMetadata) User() string {
+	return "testuser"
+}
+
+func (m mockConnMetadata) SessionID() []byte {
+	return nil
+}
+
+func (m mockConnMetadata) ClientVersion() []byte {
+	return nil
+}
+
+func (m mockConnMetadata) ServerVersion() []byte {
+	return nil
+}
+
+func (m mockConnMetadata) LocalAddr() net.Addr {
+	return &net.TCPAddr{
+		IP:   net.ParseIP("1.2.3.4"),
+		Port: 22,
+	}
+}
+
+func (m mockConnMetadata) RemoteAddr() net.Addr {
+	return &net.TCPAddr{
+		IP:   net.ParseIP("127.0.0.1"),
+		Port: 9001,
+	}
+}
+
+func TestRBAC(t *testing.T) {
+	tests := []struct {
+		name                  string
+		component             string
+		nodeExists            bool
+		openSSHNode           bool
+		shouldAuthWithRBAC    bool
+		shouldAuthWithoutRBAC bool
+	}{
+		{
+			name:                  "teleport node, regular server",
+			component:             teleport.ComponentNode,
+			nodeExists:            true,
+			openSSHNode:           false,
+			shouldAuthWithRBAC:    true,
+			shouldAuthWithoutRBAC: false,
+		},
+		{
+			name:                  "teleport node, forwarding server",
+			component:             teleport.ComponentForwardingNode,
+			nodeExists:            true,
+			openSSHNode:           false,
+			shouldAuthWithRBAC:    false,
+			shouldAuthWithoutRBAC: true,
+		},
+		{
+			name:                  "registered openssh node, forwarding server",
+			component:             teleport.ComponentForwardingNode,
+			nodeExists:            true,
+			openSSHNode:           true,
+			shouldAuthWithRBAC:    true,
+			shouldAuthWithoutRBAC: false,
```

Comment:
Do we need two booleans for that? Maybe
```
shouldAuthWithRBAC: [true/false]
```
is enough?

Reply:
I want to ensure either `canLoginWithoutRBAC` or `canLoginWithRBAC` is called, and I need two booleans to do that.